go/kbfs/libkbfs/md_ops.go

```diff
@@ -375,6 +376,55 @@ func (md *MDOpsStandard) checkRevisionCameBeforeMerkle(
 	verifyingKey kbfscrypto.VerifyingKey, irmd ImmutableRootMetadata,
 	root keybase1.MerkleRootV2, timeToCheck time.Time) (err error) {
 	ctx = context.WithValue(ctx, ctxMDOpsSkipKeyVerification, struct{}{})
+	// This gives us the KBFS merkle root that comes after the KBFS merkle
+	// root right after the one included by `root`. In other words, if the
+	// KBFS merkle root included in `root` is K0, the returned KBFS merkle
+	// root `kbfsRoot` here would be K2.
+	//
+	// Since `root` is the merkle root included by the revoke signature,
+	// anything written by the revoked device after K2 is not legit. The reason
+	// why we don't check K1 is that when K1 was built, it's possible that it
+	// only included writes happening right before the revoke happened. For
+	// example:
+	//
+	//    - 8:00am mdmerkle starts to run
+	//    - 8:01am mdmerkle scans TLF /keybase/private/alice
+	//    - 8:02am Alice's device "Playstation 4" writes to her TLF
+	//        /keybase/private/alice
+	//    - 8:03am Alice revokes her device "Playstation 4", referencing a
+	//        global merkle root that includes KBFS merkle K0
+	//    - 8:07am mdmerkle finishes building the merkle tree and publish K1,
+	//        which includes the head MD from /keybase/private/alice from 8:01,
+	//        the one before the write happened at 8:02am.
+	//    - 9:00am mdmerkle starts to run again, producing a KBFS merkle tree
+	//        K2.
+	//
+	// Looking at the per TLF metadata chain, the writes happened at 8:02
+	// happened after the MD included in K1, but it was before the revoke and
+	// was legit. So we have to compare with K2 instead of K1.
+	//
+	// To protect against a malicious server indefinitely halting building
+	// KBFS merkle trees with the purpose of slipping in malicious writes
+	// from a revoked device, we have a "contract" with the mdserver that
+	// any writes happening after a grace period of
+	// `maxAllowedMerkleGapServer` after the last KBFS merkle tree was
+	// published should be rejected. In other words, KBFS should become
+	// readonly if `maxAllowedMerkleGapServer` has passed since laste  KBFS
+	// merkle tree was published. We add a small error window to it, and
+	// enforce in client that any MD with a gap greater than
+	// `maxAllowedMerkleGap` since last KBFS merkle tree is illegal.
+	//
+	// However there's no way get a trustable timestamp for the metadata
+	// update happening at 8:02am when we haven't verified that metadata yet,
+	// we have to proxy it to some sort of ordering from merkle trees. Here we
+	// are using K2 as that proxy. Since we already need to guarantee the
+	// metadata in question has to happen before K2, we just need to make sure
+	// K2 happens within `maxAllowedMerkleGap` since the revoke happened.
+	//
+	// Unfortunately this means if mdmerkle takes longer to build, and there
+	// happens to be a revoke and some writes happening during the gap, there's
```

Comment:
This could be a little more clear/specific I think, maybe something like:

Unfortunately this means if mdmerkle takes longer than \`maxAllowedMerkleGap\` to build, and a device both writes to a TLF and then gets revoked within that gap (and those writes are the most recent writes to the TLF), there's no way for us to know for sure the write is legit, from merkle tree perspective, and future reads of the TLF will fail without manual intervention.

Reply:
It's actually two trees within the gap, so I'm changing your first sentence into "Unfortunately this means if mdmerkle takes longer than `maxAllowedMerkleGap` to build _two trees_, and a device ..."